Sources/Common/Core/DataArray/index.js

```diff
@@ -313,44 +318,38 @@ function vtkDataArray(publicAPI, model) {
 // Object factory
 // ----------------------------------------------------------------------------
 
-const DEFAULT_VALUES = {
-  name: '',
-  numberOfComponents: 1,
-  size: 0,
-  dataType: DefaultDataType,
-  rangeTuple: [0, 0],
-  // values: null,
-  // ranges: null,
-};
+function defaultValues(initialValues) {
+  return {
+    name: '',
+    numberOfComponents: 1,
+    size: 0,
+    dataType: DefaultDataType,
+    rangeTuple: [0, 0],
+    // values: null,
+    // ranges: null,
+    ...initialValues,
+  };
+}
 
 // ----------------------------------------------------------------------------
 
 export function extend(publicAPI, model, initialValues = {}) {
-  Object.assign(model, DEFAULT_VALUES, initialValues);
-
-  if (!model.empty && !model.values && !model.size) {
-    throw new TypeError(
-      'Cannot create vtkDataArray object without: size > 0, values'
-    );
-  }
-
-  if (!model.values) {
-    model.values = macro.newTypedArray(model.dataType, model.size);
-  } else if (Array.isArray(model.values)) {
-    model.values = macro.newTypedArrayFrom(model.dataType, model.values);
-  }
-
-  if (model.values) {
-    model.size = model.values.length;
-    model.dataType = getDataType(model.values);
-  }
+  Object.assign(model, defaultValues(model));
 
   // Object methods
   macro.obj(publicAPI, model);
   macro.set(publicAPI, model, ['name', 'numberOfComponents']);
 
+  const type = initialValues.dataType ? initialValues.dataType : model.dataType;
+  if (!initialValues.values) {
+    initialValues.values = macro.newTypedArray(type, model.size);
+  } else if (Array.isArray(initialValues.values)) {
+    initialValues.values = macro.newTypedArrayFrom(type, initialValues.values);
+  }
+
   // Object specific methods
   vtkDataArray(publicAPI, model);
+  publicAPI.setData(initialValues.values, initialValues.numberOfComponents);
```

Comment:
it looks like it makes sense to create a test that tests ALL the different ways to create a DataArray